core/desugarDatatypes.ml

```diff
@@ -41,6 +41,71 @@ let concrete_subkind =
   | Some subkind -> subkind
   | None         -> default_subkind
 
+(** Replaces all placeholder variables with a fresh one. *)
+class basic_freshener = object
+  inherit SugarTraversals.map as super
+
+  method! known_type_variable =
+    let module SC = SugarConstructors.SugartypesPositions in
+    function
+    | (("_" | "_anon"), None, freedom) ->
+       SC.fresh_known_type_variable freedom
+       |> super#known_type_variable
+    | v ->  super#known_type_variable v
+end
+
+let freshen_vars =
+  let module SC = SugarConstructors.SugartypesPositions in
+  (* Determine if this is an "anonymous" effect type variable, and so introduced
+     within a simple arrow (->, ~>, -@, ~>) *)
+  let is_anon_effect = function
+    | Datatype.Open ("_anon", None, `Rigid) -> true
+    | _ -> false in
+  (* If this function type is exclusively composed of anonymous effect type
+     variables. Or rather, there are no explicitly mentioned effect variables. *)
+  let all_anon_effects = object
+    inherit SugarTraversals.fold as super
+
+    val all_anon = true
+    method all_anon = all_anon
+
+    method! datatypenode = let open Datatype in
+      function
+      | Function (_, (_, eff_var), _) | Lolli (_, (_, eff_var), _)
+           when not (is_anon_effect eff_var) -> {<all_anon = false>}
+      | ty -> super#datatypenode ty
+  end in
+
+  let basic_refresh = new basic_freshener
+  and shared_refresh var = object
+    inherit basic_freshener as super
+    method! datatypenode = let open Datatype in
+      function
+      | Function (args, (effects, eff_var), ret) when is_anon_effect eff_var
+        -> super#datatypenode (Function (args, (effects, var), ret))
+      | Lolli (args, (effects, eff_var), ret) when is_anon_effect eff_var
+        -> super#datatypenode (Lolli (args, (effects, var), ret))
+      | ty -> super#datatypenode ty
+  end in
+
+  object
+    inherit basic_freshener as super
+    method! datatypenode = let open Datatype in
```

Comment:
I think you can implement this by overloading the `row` method. Then this rewrite should apply if the (sub)kind of the row variable is `Eff`. This might even make it possible to extend this syntactic sugar to user definable datatypes such as `Comp`.